src/main/kotlin/cacophony/codegen/instructions/CacophonyInstructionCovering.kt

```diff
@@ -22,18 +22,29 @@ class CacophonyInstructionCovering(private val instructionMatcher: InstructionMa
 
     private fun coverWithInstructionsForValue(node: CFGNode, register: Register): List<Instruction> {
         val matches = instructionMatcher.findMatchesForValue(node, register)
-        val bestMatch = matches.maxByOrNull { match -> match.size } ?: error("No match found for $node, ${node.javaClass}")
+        val bestMatch =
+            matches.maxByOrNull { match -> match.size } ?: error("No match with instructions for value found for $node, ${node.javaClass}")
         return coverGivenMatch(bestMatch)
     }
 
     override fun coverWithInstructions(node: CFGNode): List<Instruction> {
         val matches =
-            instructionMatcher.findMatchesForSideEffects(node)
+            instructionMatcher
+                .findMatchesForSideEffects(node)
                 .union(instructionMatcher.findMatchesForValue(node, Register.VirtualRegister()))
         val bestMatch = matches.maxByOrNull { match -> match.size } ?: error("No match found for $node, ${node.javaClass}")
         return coverGivenMatch(bestMatch)
     }
 
+    override fun coverWithInstructionsWithoutVirtualRegisters(node: CFGNode): List<Instruction> {
+        val matches = instructionMatcher.findMatchesForSideEffects(node)
+        val bestMatch =
+            matches
+                .filter { it.toFill.isEmpty() }
```

Comment:
I kinda don't like it. toFill is not empty if a pattern has valueSlot, so we check patterns that we know we'll just yank out in this filter. It's not problematic, but maybe you'll come up with better solution.

Reply:
we could create sub-interface for patterns `NoTemporaryRegistersPattern`, but I'm not sure which patterns can be there?

Or maybe broader question, because I don't know this logic too good - when exactly does a match not contain any temporary virtual registers in it?

Reply:
A pattern can have 3 types of slots: valueSlot, constantSlot and registerSlot. If it has valueSlot it means that it can accept any sort of value and this value may be result of some other computations. And suppose this result is stored in some register for example, we cannot directly use this register, because we may have second children and their computation will change this register, so we have to save it to some virtualRegister. So short answer is that if pattern has any valueSlot, then it will create virtualRegister for result from subtree.

But it could probably be optimized, if result of some subtree is in virtualRegister, we can just use it instead of creating a new one. Maybe after every match that computes something we have virtualRegister and therefore we can **always** reuse it later instead of creating new one. I'm not the author of patterns nor instructionCovering :[ 

You can look into it, or ask some other people.

Reply:
thanks for clarification, I created new interface

```
interface NoTemporaryRegistersPattern : SideEffectPattern
```

and dedicated methods in covering and matcher